viam-cartographer.go

```diff
@@ -97,46 +93,108 @@ func init() {
-	cartoSvc *cartographerService,
-) error {
-	if !cartoSvc.useLiveData {
-		return nil
+	deps registry.Dependencies,
+	config config.Service,
+	logger golog.Logger,
+	bufferSLAMProcessLogs bool,
+	executableName string,
+	sensorValidationMaxTimeoutSec int,
+	sensorValidationIntervalSec int,
+	dialMaxTimeoutSec int,
+) (slam.Service, error) {
+	ctx, span := trace.StartSpan(ctx, "viamcartographer::slamService::New")
+	defer span.End()
+
+	svcConfig, ok := config.ConvertedAttributes.(*slamConfig.AttrConfig)
+	if !ok {
+		return nil, rdkutils.NewUnexpectedTypeError(svcConfig, config.ConvertedAttributes)
 	}
 
-	var err error
-	var path string
-	paths := make([]string, 0, 1)
-	startTime := time.Now()
+	subAlgo := SubAlgo(svcConfig.ConfigParams["mode"])
+	if subAlgo != Dim2d {
+		return nil, errors.Errorf("%v does not have a 'mode: %v'",
+			string(config.Model.Name), svcConfig.ConfigParams["mode"])
+	}
 
-	for {
-		path, err = cartoSvc.getAndSaveData(ctx, cams)
-		paths = append(paths, path)
+	// Set up the data directories
+	if err := slamConfig.SetupDirectories(svcConfig.DataDirectory, logger); err != nil {
+		return nil, err
+	}
 
-		if err == nil {
-			break
-		}
+	port, dataRateMsec, mapRateSec, useLiveData, deleteProcessedData, err := slamConfig.GetOptionalParameters(
+		svcConfig,
+		localhost0,
+		defaultDataRateMsec,
+		defaultMapRateSec,
+		logger,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the lidar for the Dim2D cartographer sub algorithm
+	lidar, err := dim2d.NewLidar(ctx, deps, svcConfig.Sensors, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	cancelCtx, cancelFunc := context.WithCancel(ctx)
 
-		// This takes about 5 seconds, so the timeout should be sufficient.
-		if time.Since(startTime) >= time.Duration(cameraValidationMaxTimeoutSec)*time.Second {
-			return errors.Wrap(err, "error getting data from sensor")
+	// Cartographer SLAM Service Object
+	cartoSvc := &cartographerService{
+		primarySensorName:     lidar.Name,
+		executableName:        executableName,
+		subAlgo:               subAlgo,
+		slamProcess:           pexec.NewProcessManager(logger),
+		configParams:          svcConfig.ConfigParams,
+		dataDirectory:         svcConfig.DataDirectory,
+		useLiveData:           useLiveData,
+		deleteProcessedData:   deleteProcessedData,
+		port:                  port,
+		dataRateMs:            dataRateMsec,
+		mapRateSec:            mapRateSec,
+		cancelFunc:            cancelFunc,
+		logger:                logger,
+		bufferSLAMProcessLogs: bufferSLAMProcessLogs,
+		dev:                   svcConfig.Dev,
+	}
+
+	var success bool
+	defer func() {
+		if !success {
+			if err := cartoSvc.Close(); err != nil {
+				logger.Errorw("error closing out after error", "error", err)
+			}
 		}
-		if !goutils.SelectContextOrWait(ctx, cameraValidationIntervalSec*time.Second) {
-			return ctx.Err()
+	}()
+
+	if cartoSvc.useLiveData {
+		if err := dim2d.ValidateGetAndSaveData(cancelCtx, cartoSvc.dataDirectory, lidar,
```

Comment:
What happens here if 'sensors' was empty? We haven't checked that sensors must be non-empty if useLiveData is true.

Reply:
The evaluation of `useLiveData` ensures that the `sensors` array must contain a sensor when `use_live_data: true`: https://github.com/viamrobotics/slam/blob/519072a67f6a77476f0c32d2c524b42b75101ad1/config/config.go#L39-L41